http://sourceware.org/bugzilla/show_bug.cgi?id=9864

--- origsrc/xinit-1.1.1/startx.cpp	2008-12-17 14:59:54.000000000 -0600
+++ src/xinit-1.1.1/startx.cpp	2009-02-18 11:18:37.902875000 -0600
@@ -279,7 +279,7 @@ if [ x"$enable_xauth" = x1 ] ; then
     xauth -q -f "$xserverauthfile" << EOF
 add :$dummy . $mcookie
 EOF
-#ifdef __APPLE__
+#if defined(__APPLE__) || defined(__CYGWIN__)
     serverargs=${serverargs}" -auth '"${xserverauthfile}"'"
 #else
     serverargs=${serverargs}" -auth "${xserverauthfile}
@@ -313,7 +313,7 @@ else
 fi
 #else
 
-#ifdef __APPLE__
+#if defined(__APPLE__) || defined(__CYGWIN__)
 eval XINIT \"$client\" $clientargs -- \"$server\" $display $serverargs
 #else
 XINIT "$client" $clientargs -- "$server" $display $serverargs
