--- origsrc/xinit-1.3.4/Xsession	2014-11-27 13:40:33.968868200 -0600
+++ src/xinit-1.3.4/Xsession	2014-11-27 13:57:39.496593600 -0600
@@ -42,10 +42,8 @@ SWITCHDESKPATH=/usr/share/switchdesk
 XCLIENTS_D=/etc/X11/xinit/Xclients.d
 if [ "$#" -eq 1 ] && [ -x "$XCLIENTS_D/Xclients.$1.sh" ]; then
     exec -l $SHELL -c "$CK_XINIT_SESSION $SSH_AGENT $XCLIENTS_D/Xclients.$1.sh"
-else
 # now, we see if xdm/gdm/kdm has asked for a specific environment
-case $# in
-1)
+elif [ "$#" -ge 1 ]; then
     if [ -x "$SWITCHDESKPATH/Xclients.$1" ]; then
        exec -l $SHELL -c "$SWITCHDESKPATH/Xclients.$1";
     fi;
@@ -72,11 +70,10 @@ case $# in
        # provides 'failsafe', 'default' or 'custom'.  KDM will do the
        # same at some point
 	    if [ "$1" != "default" -a "$1" != "custom" ]; then
-		exec $CK_XINIT_SESSION $SSH_AGENT /bin/sh -c "exec -l $SHELL -c \"$1\""
+		exec $CK_XINIT_SESSION $SSH_AGENT /bin/sh -c "exec -l $SHELL -c \"$*\""
 	    fi
 	    ;;
     esac
-esac
 fi
 
 # otherwise, take default action
